src/crimson/osd/shard_services.cc

```diff
@@ -802,15 +802,19 @@ seastar::future<> ShardServices::dispatch_context_messages(
 
 seastar::future<> ShardServices::dispatch_context(
   crimson::os::CollectionRef col,
-  PeeringCtx &&ctx)
-{
-  ceph_assert(col || ctx.transaction.empty());
-  return seastar::when_all_succeed(
-    dispatch_context_messages(
-      BufferedRecoveryMessages{ctx}),
-    col ? dispatch_context_transaction(col, ctx) : seastar::now()
-  ).then_unpack([] {
-    return seastar::now();
+  PeeringCtx &&pctx)
+{
+  return seastar::do_with(
+    std::move(pctx),
```

Comment:
lgtm! Got it now, the passed `ctx` instance is local (to dispatch_context) and should be kept alive as we use it inside the coroutine.